### PR TITLE
Start of a framework for checking data files

### PR DESCRIPTION
**Pull request by takluyver**
See #31. I'm not totally happy with the design of this, but it's a starting point to experiment with.

It can be used at the command line like this:

```
$ python3 -m karabo_data.validation /gpfs/exfel/exp/FXE/201701/p002026/proc/r0078
Checking run directory: /gpfs/exfel/exp/FXE/201701/p002026/proc/r0078
Validation failed!

Index referring to data (9300) outside dataset (2480)
  dataset: INSTRUMENT/FXE_DET_LPD1M-1/DET/3CH0:xtdf/image/length
  file: /gpfs/exfel/exp/FXE/201701/p002026/proc/r0078/CORR-R0078-LPD03-S00003.h5

Index referring to data (9300) outside dataset (2480)
  dataset: INSTRUMENT/FXE_DET_LPD1M-1/DET/3CH0:xtdf/image/trainId
  file: /gpfs/exfel/exp/FXE/201701/p002026/proc/r0078/CORR-R0078-LPD03-S00003.h5
...
```

### CODE REVIEW

File: karabo_data/validation.py [hunk 1]
```diff
@@ -30,8 +31,36 @@ def validate(self):
...
+        self.check_trainids()
         return self.problems
 
+    def record(self, msg, **kwargs):
+        self.problems.append(dict(
+            msg=msg, file=self.filename, **kwargs
+        ))
+
+    def check_trainids(self):
+        ds_path = 'INDEX/trainId'
```
Comment (tmichela) [inline karabo_data/validation.py]: now that I think about it, I believe that `INDEX/trainId` should not contain zeroes at all, the ones that can contain zeros in datasets are in `INSTRUMENTS/...` but they don't always contain this field. Here we can maybe simply check if there are zeroes at all.

File: karabo_data/validation.py [hunk 2]
```diff
@@ -30,8 +31,36 @@ def validate(self):
...
+                self.record(
+                    'Zeroes in trainId index before last train ID',
+                    dataset=ds_path,
+                )
+            nonzero_tids = train_ids[train_ids != 0]
+        else:
+            nonzero_tids = train_ids
+
+        if len(nonzero_tids) > 1:
+            if not (nonzero_tids[1:] > nonzero_tids[:-1]).all():
```
Comment (tmichela) [inline karabo_data/validation.py]: I would assume that they need to increase monotonically (by 1) but we better check with IT.

File: karabo_data/validation.py (hunk 1 above)
Comment (takluyver) [reply karabo_data/validation.py]: At least some of the older datasets have the train IDs written in chunks, so there are zeroes at the end. E.g. `/gpfs/exfel/exp/SPB/201730/p900008/raw/r0273/RAW-R0273-DA01-S00000.h5`. It seems like the newer files don't do this, but it's not clear when that changed or if it's a feature we can rely on. `karabo_data` is also built to handle trailing 0s there.

**Comment by tmichela** [comment]
is this feature ready?

**Comment by takluyver** [comment]
I think it's a reasonable framework to start with. We can add more checks to it as we go.

I had a vague idea for a system where each possible problem would be a separate class, with an idenfitier you could pass to disable it along with any sub-checks. But I couldn't work out how it would all fit together, and I suspect it may be more complexity than we need. So I think the framework that we have in this PR is a reasonable starting point.

**Comment by takluyver** [comment]
Thanks!

**Pull request merged**